src/pip/_internal/cache.py

```diff
@@ -4,28 +4,41 @@
 # The following comment should be removed at some point in the future.
 # mypy: strict-optional=False
 
-import errno
 import hashlib
 import logging
 import os
+import sys
 
 from pip._vendor.packaging.utils import canonicalize_name
 
 from pip._internal.models.link import Link
 from pip._internal.utils.compat import expanduser
+from pip._internal.utils.misc import interpreter_name
 from pip._internal.utils.temp_dir import TempDirectory
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 from pip._internal.utils.urls import path_to_url
 from pip._internal.wheel import InvalidWheelFilename, Wheel
 
 if MYPY_CHECK_RUNNING:
-    from typing import Optional, Set, List, Any
+    from typing import Optional, Set, List, Any, Dict
     from pip._internal.models.format_control import FormatControl
     from pip._internal.pep425tags import Pep425Tag
 
 logger = logging.getLogger(__name__)
 
 
+def _hash_dict(d):
+    # type: (Dict[str, str]) -> str
+    """Return a sha224 of a dictionary where keys and values are strings."""
+    h = hashlib.new('sha224')
+    for k in sorted(d.keys()):
+        h.update(k.encode())
+        h.update("=".encode())
```

Comment:
I don't think it will necessarily get hit with our current usage, but this could be ambiguous if we had e.g. `{"k=v1": "v2"}` and `{"k": "v1=v2"}`. It would be nice to not have to think about that. Is there any issue with something like `json.dumps(s, sort_keys=True, separators=(",", ":"))` (wrapped in a separate function that handles the 2/3 encoding differences)?

Reply:
@chrahunt like this? f3e5d51

Reply:
Alternatively, using `\0` instead of `=` would work too. It's a bit more efficient, less flexible and maybe more under our control than json (although admittedly the probability of `json.dumps` changing output format is very low).

Reply:
Both approaches seem fine to me.
I've also seen `b'\x1D'`, `b'\x1E'` & `b'\x1F'` (respectively group, record and unit separator in ASCII) used for this kind of hashing.

Reply:
@chrahunt what do you think here? Is the last commit what you expected?

Reply:
Yes, exactly.